cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/batch/preview/LimitingInputFormat.java

```diff
@@ -34,25 +44,100 @@
  * @param <V> type of value to read
  */
 public class LimitingInputFormat<K, V> extends InputFormat<K, V> implements Configurable {
-  public static final String DELEGATE_CLASS_NAME = "io.cdap.pipeline.preview.input.classname";
-  public static final String MAX_RECORDS = "io.cdap.pipeline.preview.max.records";
+
+  static final String DELEGATE_CLASS_NAME = "io.cdap.pipeline.preview.input.classname";
+  static final String MAX_RECORDS = "io.cdap.pipeline.preview.max.records";
 
   private InputFormat<K, V> delegateFormat;
   private Configuration conf;
 
   @Override
   public List<InputSplit> getSplits(JobContext context) throws IOException, InterruptedException {
     Configuration conf = context.getConfiguration();
-    return createDelegate(conf).getSplits(context);
+    int maxRecords = conf.getInt(MAX_RECORDS, 100);
+
+    List<InputSplit> splits = createDelegate(conf).getSplits(context);
+    if (splits.size() <= 1) {
+      int limit = maxRecords;
+      return splits.stream().map(s -> new LimitingInputSplit(conf, s, limit)).collect(Collectors.toList());
+    }
+
+    // If there are more than one splits, try to read records from each split to determine what's the actual split
+    // limit per split.
+    Map<InputSplit, Integer> recordLimits = new IdentityHashMap<>();
+    TaskID taskId = new TaskID(context.getJobID(), TaskType.JOB_SETUP, 0);
+    TaskAttemptContext taskContext = new TaskAttemptContextImpl(conf, new TaskAttemptID(taskId, 0));
+    List<InputSplit> activeSplits = new LinkedList<>(splits);
+    while (maxRecords > 0 && !activeSplits.isEmpty()) {
+      int recordPerSplit = Math.max(1, maxRecords / activeSplits.size());
+
+      Iterator<InputSplit> iterator = activeSplits.iterator();
+      while (iterator.hasNext()) {
+        InputSplit split = iterator.next();
+
+        // We close the record reader in each iteration to avoid keeping all of them open to preserve memory
+        // in case there are a lot of partitions.
+        try (RecordReader<K, V> reader = createRecordReader(split, taskContext)) {
```

Comment:
instead of reading data here, another possibility is to create one combined split that contains all the original splits, and have the record reader iterate through the splits later. 

Seems like it would be better, since the data is only read once, and protects against weird behavior in certain types of input formats (for example, if there's a pubsub source where the message is acked in the reader).

Reply:
hum.. we could and move this logic to the new reader.